### PR TITLE
Adding support for tag constraints when adding a package repo with kctrl

### LINKED ISSUE #1455: Add support for tag constraints when adding a package repo with kctrl
**Describe the problem/challenge you have**

I am trying to add a package repo using `kctrl package repository add` and I would like to make use of the `.spec.fetch.imgpkgBundle.tagSelection.semver.constraints` field available on the PackageRepository CRD.

**Describe the solution you'd like**

I would like to be able to do something like `kctrl package repository add my-repo --url my-repo-without-passing-in-a-tag --semver-tag-constraints "1.8.x"` which would generate something like this:

```yaml
apiVersion: packaging.carvel.dev/v1alpha1
kind: PackageRepository
metadata:
  name: my-repo
spec:
  fetch:
    imgpkgBundle:
      image: my-repo-without-passing-in-a-tag
      tagSelection:
        semver:
          constraints: "1.8.x"
```

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by sethiyash**
#### What this PR does / why we need it:
- Support tag constraints when adding a package repo

Note: We are only supporting semver when the tag is not present in URL

#### Which issue(s) this PR fixes:

Fixes #1455

#### Does this PR introduce a user-facing change?

```release-note

```

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go [hunk 1]
```diff
@@ -325,7 +330,9 @@ func (o AddOrUpdateOptions) dryRun() error {
 
 	if tag == "" {
 		packageRepo.Spec.Fetch.ImgpkgBundle.TagSelection = &versions.VersionSelection{
-			Semver: &versions.VersionSelectionSemver{},
+			Semver: &versions.VersionSelectionSemver{
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go:333]: It seems like code is not ensuring only 1 of the semver or tag in the URL needs to be provided. We can update in the description of the flag which one takes precedence. 

Did you test what happens when both tag and semver are provided. How does Package Repository looks like?

Comment (sethiyash) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:333]: @rohitagg2020 I have added in description that tag will get precedence and added a new test case having both tag and semver

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go [hunk 2]
```diff
@@ -76,6 +77,7 @@ func NewAddCmd(o *AddOrUpdateOptions, flagsFactory cmdcore.FlagsFactory) *cobra.
 
 	// TODO consider how to support other repository types
 	cmd.Flags().StringVar(&o.URL, "url", "", "OCI registry url for package repository bundle (required)")
+	cmd.Flags().StringVar(&o.SemverTagConstraints, "semver-tag-constraints", "", "Mention tag/semver constraint when tag is not present in URL separated by ':' (tag always get precedence)")
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]:
```suggestion
	cmd.Flags().StringVar(&o.SemverTagConstraints, "semver-tag-constraints", "", "Mention tag/semver constraint when tag is not present in URL ( If both tags and semver are present, then tag gets precedence)")
```

File: .github/workflows/test-kctrl-gh.yml [hunk 3]
```diff
@@ -34,6 +34,12 @@ jobs:
       with:
         only: ytt, kbld, imgpkg, vendir, kapp
         token: ${{ secrets.GITHUB_TOKEN }}
+    - name: Login to GitHub Container Registry
```
Comment (praveenrewar) [inline .github/workflows/test-kctrl-gh.yml]: Is this required? The image is public so I think that login shouldn't be required.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go [hunk 4]
```diff
@@ -76,6 +77,7 @@ func NewAddCmd(o *AddOrUpdateOptions, flagsFactory cmdcore.FlagsFactory) *cobra.
 
 	// TODO consider how to support other repository types
 	cmd.Flags().StringVar(&o.URL, "url", "", "OCI registry url for package repository bundle (required)")
+	cmd.Flags().StringVar(&o.SemverTagConstraints, "semver-tag-constraints", "", "Mention tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go:80]:
```suggestion
	cmd.Flags().StringVar(&o.SemverTagConstraints, "semver-tag-constraints", "", "tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go [hunk 5]
```diff
@@ -116,6 +118,7 @@ func NewUpdateCmd(o *AddOrUpdateOptions, flagsFactory cmdcore.FlagsFactory) *cob
 	}
 
 	cmd.Flags().StringVarP(&o.URL, "url", "", "", "OCI registry url for package repository bundle (required)")
+	cmd.Flags().StringVarP(&o.SemverTagConstraints, "semver-tag-constraints", "", "", "Mention tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go:121]:
```suggestion
	cmd.Flags().StringVarP(&o.SemverTagConstraints, "semver-tag-constraints", "", "", "tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go (hunk 4 above)
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]:
```suggestion
	cmd.Flags().StringVar(&o.SemverTagConstraints, "semver-tag-constraints", "", "tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go (hunk 5 above)
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]:
```suggestion
	cmd.Flags().StringVarP(&o.SemverTagConstraints, "semver-tag-constraints", "", "", "tag/semver constraint when tag is not present in URL (If both tags and semver are present, then tag gets precedence)")
```

**Review by rohitagg2020 — APPROVED** [review]

File: cli/test/e2e/package_repository_test.go [hunk 6]
```diff
@@ -193,3 +190,53 @@ func TestPackageRepository(t *testing.T) {
...
+	pkgrURL := `ghcr.io/carvel-dev/kc-e2e-test-repo`
+
+	kind := "PackageRepository"
+
+	cleanUp := func() {
+		RemoveClusterResource(t, kind, pkgrName, env.Namespace, kubectl)
+	}
+
+	cleanUp()
+	//defer cleanUp()
```
Comment (praveenrewar) [inline cli/test/e2e/package_repository_test.go]: Remove comment.

**Review by praveenrewar — APPROVED** [review]
LGTM!

**Pull request merged**